test-case/check-alsabat.sh

```diff
@@ -72,11 +79,83 @@ function __upload_wav_file
     done
 }
 
+set_pga_to_unity_gain()
+{
+    tmp=$(amixer controls | grep "$1" | grep Volume)
+    search="name="
+    cname=${tmp#*$search}
+
+    # Get volume min and step to compute value for cset
+    # for 0 dB gain. The amixer line looks like
+    # "| dBscale-min=-50.00dB,step=1.00dB,mute=1"
+    scale=$(amixer cget name="$cname" | grep "dBscale" || true)
+    search="dBscale-min="
+    tmp=${scale#*$search}
+    min_db="${tmp%%dB*}"
+    search="step="
+    tmp=${scale#*$search}
+    step_db="${tmp%%dB*}"
+
+    # Get multiplied by 100 values by removing decimal dot
+    min_x100="${min_db//.}"
+    step_x100="${step_db//.}"
+    val=$(printf %d "$(((-min_x100) / step_x100))")
+
+    # Apply the computed value for requested gain
+    amixer cset name="$cname" "$val"
+}
+
+set_pgas_list_to_unity_gain()
+{
+    for pga in "$@"; do
+	dlogi "Set $pga"
+	set_pga_to_unity_gain "$pga"
+    done
+}
+
+get_snd_base()
+{
+    # Converts e.g. string hw:1,0 to /dev/snd/pcmC1D0
+    # the p or c for playback or capture is appended
+    # in the calling function.
+    tmp=${1#*"hw:"}
+    ncard="${tmp%%,*}"
+    ndevice="${tmp#*,}"
+    echo "/dev/snd/pcmC${ncard}D${ndevice}"
+}
+
+get_play_snd()
+{
+    tmp=$(get_snd_base "$1")
+    echo "${tmp}"p
+}
+
+get_capture_snd()
+{
+    tmp=$(get_snd_base "$1")
+    echo "${tmp}"c
+}
+
 # check the PCMs before alsabat test
 dlogi "check the PCMs before alsabat test"
 aplay   -Dplug$pcm_p -d 1 /dev/zero -q || die "Failed to play on PCM: $pcm_p"
 arecord -Dplug$pcm_c -d 1 /dev/null -q || die "Failed to capture on PCM: $pcm_c"
 
+# Set PGAs for PCMs to 0 dB value
+test -n "$(command -v "$TPLGREADER")" ||
+    die "Command $TPLGREADER is not available."
+
+test -n "$tplg" || die "Use -t or set environment variable TPLG to current topology"
+tplg_full_path=$(func_lib_get_tplg_path "$tplg")
+dlogi "Getting playback PGA information"
+play_snd=$(get_play_snd "$pcm_p")
+PLAY_PGAS=$($TPLGREADER "$tplg_full_path" -f "snd:$play_snd" -d pga -v)
```

Comment:
`-d pga` seems to be compatible with topology v1 only, same problem as in #1068. @ranj063 confirm?

Reply:
are we talking about the string "pga" in the kcontrol name here? Then yes it is compatible with the tplg2 as well. Take a look at the gain kcontrols in the nocodec tplg https://sof-ci.01.org/linuxpr/PR4937/build2360/devicetest/index.html?model=TGLU_RVP_NOCODEC-ipc4&testcase=verify-tplg-binary

Reply:
When run on topology v1 files, `tools/sof-tplgreader.py` shows a field named `pga`. Example:
```
 ./tools/sof-tplgreader.py /lib/firmware/intel/sof-tplg/sof-apl-nocodec.tplg -d pga


pga=PGA9.0 PGA5.0;
pga=PGA6.0;
pga=PGA11.0 PGA5.0;
```
Note the blank lines stand for DMIC and other PCMs without a `pga=` field.

What would be the equivalent for topology v2? Is there a v2 compatibility issue with `sof-tplgreader.py`?

Reply:
> Take a look at the gain kcontrols in the nocodec tplg https://sof-ci.01.org/linuxpr/PR4937/build2360/devicetest/index.html?model=TGLU_RVP_NOCODEC-ipc4&testcase=verify-tplg-binary

This device has at least 15 amixer controls with "Volu.." in the name (`name=` which may or may not be truncated). That's the list of kcontrols volume tests want, correct?

That log shows a list of pipelines, 3 of them which have `cap_name=Gain Capture`.

I don't see the connection.

Reply:
@marc-hb i see your problem. Sorry I misundertood the PGA in the picture. Looks like that's the tplgtool printing that the PGA prefix for the kcontrols. 

So, yes there's no "pga" string in tplg2 kcontrol names. Infact we consciously removed the widget name from the kcontrol name as well so you have "gain" either. So the question is how to figure out which kcontrols to test?

Can we do this during tplg parsing ie when as we parse the gain widgets, save the kcontrol names for the kcontrols associated with the widget and test only those?

Reply:
> as we parse the gain widgets, save the kcontrol names for the kcontrols associated with the widget and test only those?

Sound like a plan. Let me take a look.